---
title: "codys_awesome_gs_tailseq_stuffs"
author: "Cody Ocheltree"
date: "5/23/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## WELCOME TO THE FUTURE OF RNA


# First let's load in some packages to run the code 

```{r}
# Install reticulate package (only once)
# copy and paste this in the console: install.packages(“reticulate”)

# Load reticulate package
library(reticulate)
```

# Let's use the patented JLA Quickfix Tailseq Analyzer to align our reads:

QuickFix_Tailseq_analyzer_using_alignments_variable_input_file.py

```{python}
#Needs a CSV file (Windows!) with columns containing sequence, number of unique reads, number of reads (only columns 1 and 2 are used). No header.
#Needs FASTA file with RNA sequences to align to
#Aligns reads with tRNA sequences from FASTA file
#Returns a "Taildata" file listing sequences, counts, identified gene, 3' end positions, tail length and sequence.


path = raw_input("~/Desktop/TUT2_analysis/501701")
seqfile = raw_input("~/Desktop/TUT2_analysis/tailseq_pipeline/Human_snRNA_sequences_FASTA_4_3rea_bglobin_exoALL_7sk_7sl.txt")

from glob import glob
import csv
import os

#Generates a List (tRNAseqlist) containing gene names and sequences from the snRNA FASTA file
f = open(seqfile + '.txt')
RNAseqs = f.read() #loads the tRNA FASTA file into tRNAseqs
f.close()

RNAseqlist = []

x = 1

while x > -1:
	a = RNAseqs.find('>', x-1)
	b = RNAseqs.find('\n', a)
	c = RNAseqs.find('\n', b)
	d = RNAseqs.find('>', c)
	if d == -1:
		d = len(RNAseqs)
	while RNAseqs[d-1] == '\n':
		d -= 1

	RNAseqlist.append([RNAseqs[a+1:b],RNAseqs[c+1:d].replace("\n","")])

	x = RNAseqs.find('>', c)


#Saves a csv file with tRNA gene names and sequence
f = open('RNAsequences.csv', 'w')
csv.writer(f).writerows(RNAseqlist)
f.close()




data_filenames = glob(path + '*.csv')

for datafile in data_filenames:
	fcsv = open(datafile)

	Masterlist = [["Sequence","#Unique Reads","Gene","3'end","Tail length","Tail seq"]]

	firstline = True

	for line in csv.reader(fcsv):

		#Generates the Masterlist for the Taildata file

		if firstline: #skips first line
			firstline = False
			continue


		seq = line[0][14:] #Grabs the sequence; Note drops the 14 first nucleotides

		reads = line[2] #Grabs the number of unique reads


		#Identifies the best matching gene(s) for the current read
		linenumber = -1
		topgenescore = 0
		matchgeneseq = []
		matchgenelist = []
		matchlinenumber = []
		matchposition = []
		matchlength = []

		for geneline in RNAseqlist:
			linenumber += 1
			geneseq = geneline[1]
			x = 0
			genescore = 0

			while x <= len(geneseq) - 10:
				y = 0
				m = 0
				while y < (len(seq)) and (x + y) < len(geneseq) and m == 0:
					if seq[y] == geneseq[x+y]:
						y += 1
					else:
						m += 1

				if y > genescore:
					genescore = y
					geneposition = x
				x += 1

			if genescore == topgenescore:
				matchgeneseq.append(geneseq)
				matchgenelist.append(geneline[0])
				matchlinenumber.append(linenumber)
				matchposition.append(geneposition)
				matchlength.append(genescore)
			if genescore > topgenescore:
				matchgeneseq = [geneseq]
				matchgenelist = [geneline[0]]
				matchlinenumber = [linenumber]
				matchposition = [geneposition]
				matchlength = [genescore]
				topgenescore = genescore



		#Identifies the 3'end, tail length and tail sequence
		
		g = -1
		toss = 0
		errmsg = ''

		if matchlength[0] < 10: #Requires 10 or more matches
			toss = 1
			errmsg = 'ERR(Short)'
		
		else:		

			for geneseq in matchgeneseq:
				g += 1

				if matchlength[g] < len(seq) and toss == 0:

					x = 1
					match = 0
					while (x + matchlength[g]) < len(seq):

						if (matchposition[g] + matchlength[g] + x) < len(geneseq):						
							if seq[matchlength[g] + x] == geneseq[matchposition[g] + 
										matchlength[g] + x]:
								match += 1
						x += 1

						if x > 3 and float(match)/(x-1) >= 0.75: #Tosses the read if it matches the gene 75% at any point 3n after the mismatch
							toss = 1
							errmsg = 'ERR(Mismatch)'

					if x == 3 and match == 2: #Tosses if last 2n of seq match gene after mismatch
							toss = 1
							errmsg = 'ERR(Mismatch)'

					if toss == 0 and x > 1: #Looks for deletion or insert if the tail is longer than 1n
						y = 1
						match = 0
						while (y + matchlength[g] + 1) < len(seq):

							if (matchposition[g] + matchlength[g] + y) < len(geneseq):
								if seq[matchlength[g] + y + 1] == geneseq[matchposition[g]
										+ matchlength[g] + y]:
									match += 1						
							y += 1

							if y > 3 and float(match)/(y-1) >= 0.75: #Tosses the read if it matches the gene 75% at any point 3n after the mismatch
								toss = 1
								errmsg = 'ERR(Insert)'


						y = 1
						match = 0
						while (y + matchlength[g]) < len(seq) and (matchposition[g] + 
									matchlength[g] + y + 1) < len(geneseq):
							if matchposition[g] + matchlength[g] + y + 1 < len(geneseq):
								if seq[matchlength[g] + y] == geneseq[matchposition[g] + 
											matchlength[g] + y + 1]:
									match += 1
							y += 1

							if y > 3 and float(match)/(y-1) >= 0.75: #Tosses the read if it matches the gene 75% at any point 3n after the mismatch
								toss = 1
								errmsg = 'ERR(Deletion)'


					if toss == 0 and x > matchlength[0]: #Tosses if tail is longer than sequenced part of gene
						toss = 1
						errmsg = 'ERR(Tail longer than body)'

				if matchlength[g] == len(seq):
					x = 0



		if toss == 0:
			end = matchposition[0] + matchlength[0] - len(geneseq)
			taillength = x
			tailseq = seq[matchlength[0]:]

		else:			
			matchgenelist = []
			end = errmsg
			taillength = ''
			tailseq = ''
			


		Masterlist.append([seq,reads,matchgenelist,end,taillength,tailseq])
	

	if not os.path.isdir(path + seqfile + '/'):
		os.makedirs(path + seqfile + '/')


	fcsv.close()

	savefile = path + seqfile + '/' + datafile.split('/')[-1] + '_' + seqfile + '_Taildata.csv'
	f = open(savefile, 'w')
	csv.writer(f).writerows(Masterlist)
	f.close()
```

```{r}
path
```

